An alternative implementation of general finite mixture models

dfm opened the pull request:
For a project I needed a more general Mixture* distribution allowing for component distributions from different families. Since #1413 wasn't ready for use, I implemented my own version that is somewhat more complete and consistent with the existing `MixtureSameFamily` distribution. In case it would be of interest to this project, I figured I'd open this here. If the maintainers would like such a distribution be included, I'm happy to add some tests and respond to any feedback. If it's not of interest (with a preference for #1413 or otherwise), absolutely no worries and I'm happy to close this. Thanks!!

hessammehr commented:
Hi @dfm, author of #1413 here. Not a maintainer but comparing the two implementations, I agree that yours does seems more complete and compatible with the existing same family mixture implementation. Personally, I would be in favour of proceeding with this PR instead of #1403.

fehiepsi commented on numpyro/distributions/mixtures.py:
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
+
+class MixtureGeneral(Distribution):
+    def __init__(
+        self, mixing_distribution, distributions, *, validate_args=None
```
nit: maybe `distributions` -> `component_distributions` to match MixtureSameFamily api

dfm replied:
👍

fehiepsi commented on numpyro/distributions/mixtures.py:
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
+        _check_mixing_distribution(mixing_distribution)
+
+        self._mixture_size = jnp.shape(mixing_distribution.probs)[-1]
+        if isinstance(distributions, Distribution):
```
Hmm, do we need to support mixtures of the same distributions?

dfm replied:
Fair - we handle this use case in `Mixture`.

fehiepsi commented on docs/source/distributions.rst:
```diff
@@ -540,6 +540,13 @@ MixtureSameFamily
     :show-inheritance:
     :member-order: bysource
 
+MixtureGeneral
```
Maybe also `Mixture` (not autoclass but `.. autofunction::`)

fehiepsi reviewed (commented):
Thanks @dfm! This is a nice distribution to have. Given that @hessammehr would like to close that PR (thanks @hessammehr for the initial take), I think we can move ahead with this one. :)

fehiepsi commented on numpyro/distributions/mixtures.py:
```diff
@@ -10,6 +10,21 @@
 
+def Mixture(mixing_distribution, distributions, *, validate_args=None):
+    """
+    Mixture distribution.
```
Could you add an example section for this as in `MixtureSameFamily`?

dfm commented:
Thanks for the comments @fehiepsi! I'll incorporate the comments and add some docs and tests ASAP.

dfm commented on numpyro/distributions/mixtures.py:
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
+            )
+
+        # FIXME: It would be good to check the support, but __eq__ isn't implemented
+        #        properly for most constraints...
```
Done!

dfm commented on numpyro/distributions/mixtures.py:
```diff
@@ -211,3 +223,208 @@ def log_prob(self, value, intermediates=None):
+            mixing_distribution.batch_shape,
+            *(d.batch_shape for d in distributions),
+        )
+        event_shape = lax.broadcast_shapes(
```
Good point - done.

dfm commented:
@fehiepsi — I'd addressed your comments, added some docstrings, and tests for this new distribution.

Also since you last looked I refactored much of the common implementation details for `MixtureSameFamily` and `MixtureGeneral` into a common abstract base class. This definitely simplifies the implementation without changing behavior, so I think it's probably a good choice, but always happy for more feedback!

fehiepsi reviewed (commented):
Hi @dfm, could you merge master branch to make CI pass?

dfm commented on numpyro/distributions/mixtures.py:
```diff
@@ -10,19 +10,166 @@
+        specifying the weights for each mixture components. The size of this
+        distribution specifies the number of components in the mixture,
+        ``mixture_size``.
+    :param distributions: Either a list of component distributions or a single
```
Good catch - thanks!!

dfm commented:
Pulled from `master` and tests passing locally.

fehiepsi reviewed (approved):
Wow a lot of tricky design! Thanks a lot, @dfm!

dfm commented:
@fehiepsi — Thanks for your review and all your work on this project!!
